ledger/participant-integration-api/src/main/scala/platform/store/backend/common/QueryStrategy.scala

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.backend.common
+
+import anorm.NamedParameter
+import com.daml.lf.data.Ref
+
+trait QueryStrategy {
+
+  /** An expression resulting to a boolean, to check equality between two SQL expressions
+    *
+    * @return plain SQL which fits the query template
+    */
+  def columnEqualityBoolean(column: String, value: String): String =
+    s"""$column = $value"""
+
+  /** This populates the following part of the query:
+    *   SELECT ... WHERE ... ORDER BY ... [THIS PART]
+    *
+    * @param limit optional limit
+    * @return a tuple:
+    *         - first part is a plain SQL which fits the query template
+    *         - second part is a list of NamedParameters
+    *         Important note: the NamedParameters should confirm to the generated SQL expression (all names there should match exactly)
+    *         Important note: implementor have to make sure that all used names are unique here in the scope of the whole query
+    */
+  def limitClause(limit: Option[Int]): (String, List[NamedParameter]) =
+    limit
+      .map(to => s"fetch next {tolc} rows only" -> List[NamedParameter]("tolc" -> to))
+      .getOrElse("" -> Nil)
+
+  /** An expression resulting to a boolean to check whether:
+    *   - the party set defined by columnName and
+    *   - the party set defined by parties
+    * have at least one element in common (eg their intersection is non empty).
+    *
+    * @param columnName the SQL table definition which holds the set of parties
+    * @param parties set of parties
+    * @param paramNamePostfix a postfix to prevent collision with anorm NamedParameter-s
```

Comment:
I am not sure that is practical: to be able to do this on the caller side it would involve parsing the resturned clause string and replacing names in there, and also mapping the NamedParams similarily. Or maybe I am missing your point?

Reply:
True, mapping the NamedParams is easy, but parsing the query string is not. 

How about at least making sure all functions that generate clauses (i.e., ones that return `(String, List[NamedParameter])`) are given a unique suffix? I don't like the fact that helpers like `EventStorageBackendTemplate.witnessesWhereClause` or `QueryStrategy.limitClause` have to pick a suffix that is unique within the whole query, without knowing anything about the query.

Reply:
Yeah this whole named parameters and composition is something I did not like much, so I took now a different approach to respect simplicity and related comments here.
* I was going back to the foundations with anorm, and observed with the SQL"" string interpolation we could eliminate all related issues at once
* Also noticed this is considered safer and faster approach compared to the SQL() style named parameter approach by anorm too (truly: the string interpolation is using macros + scala-string interpolation, and the named parameter is using parser combinators to rip apart strings)
* I draw a custom string interpolation layer on top of anorm, supporting SQL"" and cSQL"" interpolations where the latter results to CompositeSql-s which we can be used in the former, which behaves same as the anorm-counterpart. Simply put: this enables composition of string-interpolation style expressed anorm constructs.
* Quite happy with the results: client code got much easier, the whole naming/prefixing/collision problem is just gone, actually removed code...with the trade-off to have a non trivial string-fusing tree traversal algorithm to support nested CompositeSql-s. (if only anorm would support this by desing...)

check it out!